op-node/rollup/derive/span_channel_out.go

```diff
@@ -77,115 +114,119 @@ func (co *SpanChannelOut) AddBlock(rollupCfg *rollup.Config, block *types.Block)
 		return 0, fmt.Errorf("failed to convert SpanBatch into RawSpanBatch: %w", err)
 	}
-	// Encode RawSpanBatch into bytes
-	if err = rlp.Encode(&buf, NewBatchData(rawSpanBatch)); err != nil {
+
+	// switch to the other buffer and reset it for new use
+	// (the RLP buffer which is being made inactive holds the RLP encoded span batch just before the new batch was added)
+	co.swapRLP()
+	co.activeRLP().Reset()
+	if err = rlp.Encode(co.activeRLP(), NewBatchData(rawSpanBatch)); err != nil {
 		return 0, fmt.Errorf("failed to encode RawSpanBatch into bytes: %w", err)
 	}
-	// Ensure that the total size of all RLP elements is less than or equal to MAX_RLP_BYTES_PER_CHANNEL
-	if buf.Len() > MaxRLPBytesPerChannel {
+
+	// check the RLP length against the max
+	if co.activeRLP().Len() > MaxRLPBytesPerChannel {
 		return 0, fmt.Errorf("could not take %d bytes as replacement of channel of %d bytes, max is %d. err: %w",
-			buf.Len(), co.rlpLength, MaxRLPBytesPerChannel, ErrTooManyRLPBytes)
-	}
-	co.rlpLength = buf.Len()
-
-	// If the channel is full after this block is appended, we should use preserved data.
-	// so copy the compressed data to reader
-	if len(co.spanBatch.Batches) > 1 {
-		_, err = io.Copy(co.reader, co.compress)
-		if err != nil {
-			// Must reset reader to avoid partial output
-			co.reader.Reset()
-			return 0, fmt.Errorf("failed to copy compressed data to reader: %w", err)
-		}
+			co.activeRLP().Len(), co.inactiveRLP().Len(), MaxRLPBytesPerChannel, ErrTooManyRLPBytes)
 	}
 
-	// Reset compressor to rewrite the entire span batch
-	co.compress.Reset()
-	// Avoid using io.Copy here, because we need all or nothing
-	written, err := co.compress.Write(buf.Bytes())
-	// Always flush (for BlindCompressor to check if it's full)
-	if err := co.compress.Flush(); err != nil {
-		return 0, fmt.Errorf("failed to flush compressor: %w", err)
+	// if the compressed data *plus* the new rlp data is under the target size, return early
+	// this optimizes out cases where the compressor will obviously come in under the target size
+	rlpGrowth := co.activeRLP().Len() - co.lastCompressedRLPSize
+	if uint64(co.compressed.Len()+rlpGrowth) < co.target {
+		return uint64(co.lastCompressedRLPSize), nil
 	}
-	if co.compress.FullErr() != nil {
-		err = co.compress.FullErr()
+
+	// we must compress the data to check if we've met or exceeded the target size
+	if err = co.compress(); err != nil {
+		return 0, err
+	}
+	co.lastCompressedRLPSize = co.activeRLP().Len()
+
+	// if the channel is now full, either return the compressed data, or the compressed previous data
+	if err := co.FullErr(); err != nil {
+		// if there is only one batch in the channel, it *must* be returned
 		if len(co.spanBatch.Batches) == 1 {
-			// Do not return ErrCompressorFull for the first block in the batch
-			// In this case, reader must be empty. then the contents of compressor will be copied to reader when the channel is closed.
-			err = nil
+			return uint64(co.compressed.Len()), nil
```

Comment:
Why do we suddenly return the compressed length here, while above, we return uncompressed rlp sizes? In the old godoc of this function, it also said that it returns the rlp size. Would be good to add back a godoc on what integer is actually returned here. And would also be interesting to understand how this returned number is even used? I think it might only be relevant for singular batches...
This also hints at missing assertions of unit tests of `AddSingularBatch`.

Reply:
There's two more returns of the compressed size below, whereas before, the uncompressed size was returned.

Reply:
IIRC the channel builder discards the length output and only cares about errors, but you're right that tests might be missing here. I'll go through the len returns and make sure they line up with the old ones.

Reply:
I did some quick digging on this:

* There is *no* production code which cares about the `written` length
* There is only one test that cares about the `written`, and it wasn't testing Span Batches. I've expanded it to test both batch types.

And you are correct, I should be returning the RLP length in all cases, as it's the best representation of what's been written. It means slightly different than singular batch written, but it is pretty appropriate.

Reply:
I just put out a commit which addresses all your comments, but does not *yet* remove the `written` size.

Changing the `ChannelOut` interface to not include the written bytes mean we'll have to figure out another way to write `TestOutputFrameNoEmptyLastFrame`, which is currently the only user of the value.

Happy to do so, just avoiding it for the moment :)

Reply:
see comment in the test, I think we can remove it and just access internal state in that one test, because internal state is already accessed for the span channel out case.